### DISCUSSION TITLE
Scalars Prolongation and Restriction

### DISCUSSION THREAD

**Discussion by saisandeepdammati** [PrincetonUniversity/athena#481] [Q&A]
Howdy!

I am currently using scalar implementation in Athena++ for combustion applications with multiple species. Each primitive scalar (r_n) represents the species mass fraction (Y_n) of nth species. Then, we need to satisfy the following constraint in each computational cell:

 **sum(r_n) = sum(Y_n) = 1.0 (or) sum(s_n) = sum(rho*Y_n) = rho**

where, conserved scalar s_n = rho*Y_n i.e. species mass density of nth species.

My question is, if at time t=0, I initialize the scalars on each meshblock such that the above constraint is satisfied on a cell to cell basis, will the scalars still satisfy the above constraint after prolongation or restriction? Or do we have to enforce it manually?

Thank you very much.

**Comment by c-white** [comment]
Currently there is no enforcement of this constraint, in order to allow passive scalars to represent non-constrained variables (e.g. entropy). So this would need to be enforced manually. The most straightforward way would be to use the `EnrollUserExplicitSourceFunction` interface. The last argument to user source functions is a modifiable array of s_n values after the flux divergence update has occurred.
